ospfd/ospf_gr.c

```diff
@@ -0,0 +1,824 @@
+				 struct ospf_interface *oi, struct stream *s)
+{
+	struct grace_tlv_graceperiod tlv_period = {};
+	struct grace_tlv_restart_reason tlv_reason = {};
+	struct grace_tlv_restart_addr tlv_address = {};
+
+	/* Put grace period. */
+	tlv_period.header.type = htons(GRACE_PERIOD_TYPE);
+	tlv_period.header.length = htons(GRACE_PERIOD_LENGTH);
+	tlv_period.interval = htonl(gr_info->grace_period);
+	stream_put(s, &tlv_period, sizeof(tlv_period));
+
+	/* Put restart reason. */
+	tlv_reason.header.type = htons(RESTART_REASON_TYPE);
+	tlv_reason.header.length = htons(RESTART_REASON_LENGTH);
+	if (gr_info->restart_support)
+		tlv_reason.reason = OSPF_GR_SW_RESTART;
+	else
+		tlv_reason.reason = OSPF_GR_UNKNOWN_RESTART;
+	stream_put(s, &tlv_reason, sizeof(tlv_reason));
+
+	/* Put IP address. */
+	if (oi->type == OSPF_IFTYPE_BROADCAST || oi->type == OSPF_IFTYPE_NBMA
+	    || oi->type == OSPF_IFTYPE_POINTOMULTIPOINT) {
+		tlv_address.header.type = htons(RESTARTER_IP_ADDR_TYPE);
+		tlv_address.header.length = htons(RESTARTER_IP_ADDR_LEN);
+		tlv_address.addr = oi->address->u.prefix4;
+		stream_put(s, &tlv_address, sizeof(tlv_address));
+	}
+}
+
+/* Generate Grace-LSA for a given interface. */
+static struct ospf_lsa *ospf_gr_lsa_new(struct ospf_interface *oi)
+{
+	struct stream *s;
+	struct lsa_header *lsah;
+	struct ospf_lsa *new;
+	uint8_t options, lsa_type;
+	struct in_addr lsa_id;
+	uint32_t lsa_id_host_byte_order;
+	uint16_t length;
+
+	/* Create a stream for LSA. */
+	s = stream_new(OSPF_MAX_LSA_SIZE);
+
+	lsah = (struct lsa_header *)STREAM_DATA(s);
+
+	options = LSA_OPTIONS_GET(oi->area);
+	options |= LSA_OPTIONS_NSSA_GET(oi->area);
+	options |= OSPF_OPTION_O;
+
+	lsa_type = OSPF_OPAQUE_LINK_LSA;
+	lsa_id_host_byte_order = SET_OPAQUE_LSID(OPAQUE_TYPE_GRACE_LSA, 0);
+	lsa_id.s_addr = htonl(lsa_id_host_byte_order);
+
+	/* Set opaque-LSA header fields. */
+	lsa_header_set(s, options, lsa_type, lsa_id, oi->ospf->router_id);
+
+	/* Set opaque-LSA body fields. */
+	ospf_gr_lsa_body_set(&oi->ospf->gr_info, oi, s);
+
+	/* Set length. */
+	length = stream_get_endp(s);
+	lsah->length = htons(length);
+
+	/* Now, create an OSPF LSA instance. */
+	new = ospf_lsa_new_and_data(length);
+
+	if (IS_DEBUG_OSPF_GR)
+		zlog_debug("LSA[Type%d:%pI4]: Create an Opaque-LSA/GR instance",
+			   lsa_type, &lsa_id);
+
+	new->area = oi->area;
+	new->oi = oi;
+	SET_FLAG(new->flags, OSPF_LSA_SELF);
+	memcpy(new->data, lsah, length);
+	stream_free(s);
+
+	return new;
+}
+
+/* Originate and install Grace-LSA for a given interface. */
+static struct ospf_lsa *ospf_gr_lsa_originate(struct ospf_interface *oi)
+{
+	struct ospf_lsa *lsa, *old;
+
+	if (ospf_interface_neighbor_count(oi) == 0)
+		return NULL;
+
+	/* Create new Grace-LSA. */
+	lsa = ospf_gr_lsa_new(oi);
+	if (!lsa) {
+		zlog_warn("%s: ospf_gr_lsa_new() failed", __func__);
+		return NULL;
+	}
+
+	/* Find the old LSA and increase the seqno. */
+	old = ospf_gr_lsa_lookup(oi->ospf, oi->area);
```

Comment:
just checking here ... it seems like if we have an existing grace lsa we would modify that rather than create/install a new one? or delete/flush the old one? why are we keeping the old and new ones around at the same time?

Reply:
@riw777 sorry for the late reply, I completely missed your review comment somehow.

I agree it's confusing, but the correct way to update an LSA in ospfd is to create a new one with a higher sequence number and then install it. `ospf_lsa_install()` will then take care of discarding the old version of the same LSA, along with doing other things like updating the checksum, triggering SPF when appropriate, etc.